chain/client/src/stateless_validation/chunk_endorsement/tracker_v2.rs

```diff
@@ -0,0 +1,132 @@
+        Self {
+            epoch_manager,
+            store,
+            chunk_endorsements: Mutex::new(LruCache::new(
+                NonZeroUsize::new(NUM_CHUNKS_IN_CHUNK_ENDORSEMENTS_CACHE).unwrap(),
+            )),
+        }
+    }
+
+    // Validate the chunk endorsement and store it in the cache.
+    pub(crate) fn process_chunk_endorsement(
+        &self,
+        endorsement: ChunkEndorsementV2,
+    ) -> Result<(), Error> {
+        // Check if we have already received chunk endorsement from this validator.
+        let key = endorsement.chunk_production_key();
+        let account_id = endorsement.account_id();
+        if self
+            .chunk_endorsements
+            .lock()
+            .unwrap()
+            .peek(&key)
+            .is_some_and(|entry| entry.contains_key(account_id))
+        {
+            tracing::debug!(target: "client", ?endorsement, "Already received chunk endorsement.");
+            return Ok(());
+        }
+
+        // Validate the chunk endorsement and store it in the cache.
+        if validate_chunk_endorsement(self.epoch_manager.as_ref(), &endorsement, &self.store)? {
+            let mut chunk_endorsements = self.chunk_endorsements.lock().unwrap();
+            let entry = chunk_endorsements.get_or_insert_mut(key, || HashMap::new());
+            entry.insert(account_id.clone(), endorsement);
+        };
+        Ok(())
+    }
+
+    /// Called by block producer.
+    /// Returns ChunkEndorsementsState::Endorsed if node has enough signed stake for the chunk
+    /// represented by chunk_header.
+    /// Signatures have the same order as ordered_chunk_validators, thus ready to be included in a block as is.
+    /// Returns ChunkEndorsementsState::NotEnoughStake if chunk doesn't have enough stake.
+    pub(crate) fn compute_chunk_endorsements(
+        &self,
+        chunk_header: &ShardChunkHeader,
+    ) -> Result<ChunkEndorsementsState, Error> {
+        let shard_id = chunk_header.shard_id();
+        let epoch_id =
+            self.epoch_manager.get_epoch_id_from_prev_block(chunk_header.prev_block_hash())?;
+        let protocol_version = self.epoch_manager.get_epoch_protocol_version(&epoch_id)?;
+        if !ProtocolFeature::StatelessValidation.enabled(protocol_version) {
+            // Return an empty array of chunk endorsements for older protocol versions.
+            return Ok(ChunkEndorsementsState::Endorsed(None, vec![]));
+        }
+
+        let height_created = chunk_header.height_created();
+        let key = ChunkProductionKey { shard_id, epoch_id, height_created };
+
+        let chunk_validator_assignments = self.epoch_manager.get_chunk_validator_assignments(
+            &epoch_id,
+            chunk_header.shard_id(),
+            chunk_header.height_created(),
+        )?;
+
+        // Get the chunk_endorsements for the chunk from our cache.
+        // Note that these chunk endorsements are already validated as part of process_chunk_endorsement.
```

Comment:
nit. it feels like `process_chunk_endorsement` is called before `compute_chunk_endorsements`, is that correct? if so the naming of this function is confusing since it feels like we compute the endorsements first here, but in fact this function is collecting the already-processed endorsements and creating the final endorsement state. can we find another name, eg. finalize_chunk_endorsements?